File: GCD_Discount.py
import math

t = int(input())
for _ in range(t):
    n = int(input())
    arr = list(map(int, input().split()))
    brr = list(map(int, input().split()))
    
    prefix = [0] * n
    suffix = [0] * n
    prefix[0] = arr[0]
    suffix[n - 1] = arr[n - 1]
    
    for i in range(1, n):
        prefix[i] = math.gcd(prefix[i - 1], arr[i])
    
    for i in range(n - 2, -1, -1):
        suffix[i] = math.gcd(suffix[i + 1], arr[i])
    
    mx = prefix[n - 1]
    
    for i in range(n):
        if i == 0:
            temp = math.gcd(suffix[i + 1], brr[i])
            mx = max(mx, temp)
        elif i == n - 1:
            temp = math.gcd(prefix[i - 1], brr[i])
            mx = max(mx, temp)
        else:
            left = math.gcd(prefix[i - 1], suffix[i + 1])
            right = math.gcd(left, brr[i])
            mx = max(mx, right)
    
    print(mx)

File: GCD_Discount.cpp
#include <bits/stdc++.h>
using namespace std;

int main(){
	int t;
	cin >> t;
	while(t--) {
	    int n;
	    cin >> n;
	    vector<int> arr(n, 0);
	    vector<int> brr(n, 0);
	    for(int i = 0 ; i < n; i++){
        cin >> arr[i];
      }
	    for(int i = 0; i < n; i++){
        cin >> brr[i];
      }
	    vector<int> prefix(n);
	    vector<int> suffix(n);
	    prefix[0] = arr[0];
	    suffix[n - 1] = arr[n - 1];
	    for(int i = 1; i < n; i++){
	        prefix[i] = __gcd(prefix[i - 1], arr[i]);
	    }
	    for(int i = n - 2; i >= 0; i--) {
	        suffix[i] = __gcd(suffix[i + 1], arr[i]);
	    }
	    int mx = prefix[n - 1];
	    for(int i = 0; i < n; i++){
	        if(i == 0){
	            int temp = __gcd(suffix[i + 1], brr[i]);
	            mx = max(mx, temp);
	        }else if(i == n-1) {
	            int temp = __gcd(prefix[i - 1], brr[i]);
	            mx = max(mx, temp); 
	        }else{
	            int left = __gcd(prefix[i - 1], suffix[i + 1]);
	            int right = __gcd(left, brr[i]);
	            mx = max(mx, right);
	        }
	    }
	    cout<< mx <<endl;
	}
	return 0;
}
